analysis/src/scope_analyzer.rs

```diff
@@ -0,0 +1,91 @@
+use codespan::Span;
+use parser::{Statement, StatementKind};
+use std::collections::HashMap;
+
+#[derive(Clone, Debug)]
+pub struct Scope {
+    pub name: String,
+    pub span: Span,
+    pub children: Vec<Scope>,
+}
+
+impl Scope {
+    fn find_inner_scope(&self, index: usize) -> Option<Vec<Scope>> {
+        if index < self.span.start || index >= self.span.end {
+            return None;
+        }
+        
+        for child in &self.children {
+            if let Some(inner_scope) = child.find_inner_scope(index) {
+                return Some([&[self.clone()], &inner_scope[..]].concat());
+            }
+        }
+        
+        Some(vec![self.clone()])
+    }
+}
+
+pub struct ScopeAnalyzer {
+    pub statements: Vec<Statement>,
+    pub stack: Vec<String>,
+    pub symtab: HashMap<String, Span>,
+}
+
+impl ScopeAnalyzer {
+    pub fn new(statements: Vec<Statement>) -> Self {
+        Self {
+            statements,
+            stack: Vec::new(),
+            symtab: HashMap::new(),
+        }
+    }
+
+    pub fn analyze(&mut self) -> Vec<Scope> {
+        let statements = self.statements.clone();
+        let scopes: Vec<Scope> = statements.iter().cloned().flat_map(|statement| self.parse_statement(statement)).collect();
+
+        scopes
+    }
+
+    fn parse_statement(&mut self, statement: Statement) -> Option<Scope> {
+        match statement.kind {
+            StatementKind::Scope(Some(name), statements) => {
+                self.symtab.insert([&self.stack[..], &[name.lexeme.clone()]].concat().join("::").to_string(), name.span);
+                self.stack.push(name.lexeme.clone());
+                let scopes = statements.iter().cloned().flat_map(|stmt| self.parse_statement(stmt)).collect();
+                self.stack.pop();
+
+                Some(Scope {
+                    name: name.lexeme,
+                    children: scopes,
+                    span: statement.span,
+                })
+            }
```

Comment:
```suggestion
            }
            StatementKind::Enum(Some(name), expressions) => {
                self.symtab.insert([&self.stack[..], &[name.lexeme.clone()]].concat().join("::").to_string(), name.span);

                Some(Scope {
                    name: name.lexeme,
                    children: vec![],
                    span: statement.span,
                })
            }
```
_After_ pulling in the changes from `main`

Side note: do you prefer `vec![]` or `Vec::new()`? Do you think it's an important thing to be consistent on throughout the codebase?

Reply:
I don't have much of an opinion. My use of vec![] is mainly because of my javascript inline array literal habit